### PR TITLE
refactor backend

### PR DESCRIPTION
**Pull request by doms**
## Background

Refactored backend based off of [Building and Testing a REST API in Go with Gorilla Mux and PostgreSQL](https://semaphoreci.com/community/tutorials/building-and-testing-a-rest-api-in-go-with-gorilla-mux-and-postgresql)

oh, and since this is system design we are rolling with, this closes #1, #4, #5 

## Notes

still need to implement other stuff, but let's get this merged in first, then build off of it.

### CODE REVIEW

File: app.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+// App - ...
+type App struct {
+	Router *mux.Router
+	DB     *sql.DB
+}
+
+// Initialize - connect to DB and initialize routes
+func (a *App) Initialize(user, password, dbname string) {
+	// sslmode needs to be require on heroku, but disabled locally
+	// TODO: check env and dynamically build connection string (dev, test, prod)
```
[the remark this answers is not in the document]
Comment (kolten) [reply app.go]: [We could use this](https://github.com/joho/godotenv). Heroku will have a long URL string in the environment variables. So we could set a DATABASE_URL env var and remove the three params.

File: models/user.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+// DestroyUser - nukes a user account
+func (u *User) DestroyUser(db *sql.DB) error {
+	_, err := db.Exec("DELETE FROM users WHERE id=$1", u.ID)
+
+	return err
+}
+
+// GetUsers - returns a slice of all users
+// should only use this method for testing
```
[the remark this answers is not in the document]
Comment (kolten) [reply models/user.go]: Agreed,

File: app.go [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+	Router *mux.Router
+	DB     *sql.DB
+}
+
+// Claims - will be encoded to a JWT
+// Add jwt.StandardClaims as an embedded type, to provide fields like expiry time
+type Claims struct {
+	Email string `json:"email"`
+	jwt.StandardClaims
+}
```
Comment (doms) [inline app.go]: Struct used to construct JWT to store in cookies ✨

File: app.go [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+	}
+
+	// "public" info
+	userJSON := &models.UserJSON{
+		ID:        u.ID,
+		Name:      u.Name,
+		Email:     u.Email,
+		CreatedAt: u.CreatedAt,
+		UpdatedAt: u.UpdatedAt,
+	}
```
Comment (doms) [inline app.go]: I figured this data is safe enough to be "public", but we can discuss in further detail.

File: app.go [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+
+	// Set new token cookie
+	http.SetCookie(w, &http.Cookie{
+		Name:    "token",
+		Value:   tokenString,
+		Expires: expirationTime,
+	})
+
+	fmt.Println(w)
+}
```
Comment (doms) [inline app.go]: Oops, this was used for debugging, I'll get rid of it when I clean up the rest.

File: models/user.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+// UserJSON - safe user object
+type UserJSON struct {
+	ID        int       `json:"id"`
+	Name      string    `json:"name"`
+	Email     string    `json:"email"`
+	CreatedAt time.Time `json:"created_at"`
+	UpdatedAt time.Time `json:"updated_at"`
+}
+
```
Comment (doms) [inline models/user.go]: I figured this information is "safe" (as far as passing back and forth with minimum risk of leaks), but we can discuss in detail.

File: user_test.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"log"
+	"net/http"
+	"testing"
+)
+
```
Comment (doms) [inline user_test.go]: I moved all the user tests to its own file, and have confirmed they still work. This should allow for cleaner testing once we get around to writing tests for other components (like playlists requests and such).

File: utils/utils.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+	return string(bytes), err
+}
+
+// CheckPasswordHash - compares a password with its hash
+func CheckPasswordHash(password, hash string) error {
+	hashedPassword := []byte(hash)
+	plainTextPassword := []byte(password)
+	return bcrypt.CompareHashAndPassword(hashedPassword, plainTextPassword)
+}
+
```
Comment (doms) [inline utils/utils.go]: Still can't believe it took hours to realize the args were in the wrong order 🙄 😂

**Pull request merged**